core/example/benchmark/Cabana_CommPerformance.cpp

```diff
@@ -0,0 +1,586 @@
+            Cabana::scatter( halo_fast, s2 );
+
+            s3 = comm_particles.template slice<3>();
+            Cabana::scatter( halo_fast, s3 );
+
+            Cabana::deep_copy( particles, comm_particles );
+            halo_slice_scatter.stop(fraction);
+        }
+    }
+
+    // Output results.
+    halo_fast_create.outputResults( file, "send_bytes", comm_bytes );
+    halo_general_create.outputResults( file, "send_bytes", comm_bytes );
+    halo_aosoa_gather.outputResults( file, "send_bytes", comm_bytes );
+    halo_slice_gather.outputResults( file, "send_bytes", comm_bytes );
+    halo_slice_scatter.outputResults( file, "send_bytes", comm_bytes );
+}
+
+//---------------------------------------------------------------------------//
+// main
+int main( int argc, char* argv[] )
+{
+    // Initialize environment
+    MPI_Init( &argc, &argv );
+    Kokkos::initialize( argc, argv );
+
+    // Number of particles on this rank.
+    std::size_t num_particle = std::atoi( argv[1] );
+
+    // Get the name of the output file.
+    std::string filename = argv[2];
+
+    // Barier before continuing.
+    MPI_Barrier( MPI_COMM_WORLD );
+
+    // Get comm rank;
+    int comm_rank;
+    MPI_Comm_rank( MPI_COMM_WORLD, &comm_rank );
+
+    // Get comm size;
+    int comm_size;
+    MPI_Comm_size( MPI_COMM_WORLD, &comm_size );
+
+    // Open the output file on rank 0.
+    std::fstream file;
+    if ( 0 == comm_rank )
+        file.open( filename, std::fstream::out );
+
+    // Output problem details.
+    if ( 0 == comm_rank )
+    {
+        std::size_t total_num_p = num_particle * comm_size;
+        file << std::endl;
+        file << "Cabana Comm Performance Benchmark" << std::endl;
+        file << "----------------------------------------------" << std::endl;
+        file << "MPI Ranks: " << comm_size << std::endl;
+        file << "Particle per MPI Rank/GPU: " << num_particle << std::endl;
+        file << "Total number of particles: " << total_num_p << std::endl;
+        file << "----------------------------------------------" << std::endl;
+        file << std::endl;
+    }
+
+    // Transfer GPU data to CPU, communication on CPU, and transfer back to
+    // GPU.
+    performanceTest<Kokkos::CudaSpace,Kokkos::HostSpace>(
+        file, num_particle, "cuda_host_" );
+
+    // Do everything on the CPU.
+    performanceTest<Kokkos::HostSpace,Kokkos::HostSpace>(
+        file, num_particle, "host_host_" );
+
+    // Do everything on the GPU with regular GPU memory.
+    performanceTest<Kokkos::CudaSpace,Kokkos::CudaSpace>(
+        file, num_particle, "cuda_cuda_" );
+
+    // Do everything on the GPU with UVM GPU memory.
+    performanceTest<Kokkos::CudaUVMSpace,Kokkos::CudaUVMSpace>(
+        file, num_particle, "cudauvm_cudauvm_" );
+
+    // Close the output file on rank 0.
+    if ( 0 == comm_rank )
+        file.close();
+
+    // Barier before finalizing.
+    MPI_Barrier( MPI_COMM_WORLD );
```

Comment:
Why so?

Reply:
Seemed like a good idea but probably not necessary. I will remove.